### PR TITLE
Modular Assistive Driving System (MADS)

### PR DESCRIPTION
**Pull request by devtekve**
**No description provided yet. Please note:** This is *not* ready for public use. Currently, it may contain bugs and mismatched states that could lead to bans for users who are not actively known SP developers. **Please do not attempt to install** at this time. However, feel free to review the code and leave comments if you have questions or concerns.

- [x] Chrysler
- [x] Ford
- [x] GM
- [x] Honda
- [x] Hyundai
- [x] Mazda
- [x] Nissan
- [x] Subaru
- [x] Toyota
- [x] Volkswagen

Tesla will not be supported at the moment due to inconsistent engage/disengage signals for MADS toggling.

**TODOs for future PRs**:
- Tesla: Identify consistent signals for user explicit entry/exit

Relates to: 
* SP PR: https://github.com/sunnypilot/sunnypilot/pull/446
* Panda PR: https://github.com/sunnypilot/panda/pull/40

### CODE REVIEW

File: opendbc/car/hyundai/carcontroller.py [hunk 1]
```diff
@@ -6,7 +6,8 @@
 from opendbc.car.hyundai.carstate import CarState
 from opendbc.car.hyundai.hyundaicanfd import CanBus
 from opendbc.car.hyundai.values import HyundaiFlags, Buttons, CarControllerParams, CAR
-from opendbc.car.interfaces import CarControllerBase
+
+from openpilot.sunnypilot.selfdrive.car.hyundai.carcontroller import CarControllerSP
```
Comment (devtekve) [inline opendbc/car/hyundai/carcontroller.py:10]: Because this is defined on SP repo and not on opendbc, a reviewer would be unable to quickly see and understand what is this new controller bringing. We should keep make-specific changes together on the same repo.

see https://github.com/sunnypilot/sunnypilot/pull/446/files#r1843398527

File: opendbc/car/hyundai/hyundaican.py [hunk 2]
```diff
@@ -50,7 +51,10 @@ def create_lkas11(packer, frame, CP, apply_steer, steer_req,
     # FcwOpt_USM 2 = Green car + lanes
     # FcwOpt_USM 1 = White car + lanes
     # FcwOpt_USM 0 = No car + lanes
-    values["CF_Lkas_FcwOpt_USM"] = 2 if enabled else 1
+    if mads.enable_mads:
+      values["CF_Lkas_FcwOpt_USM"] = 2 if mads.lat_active or mads.disengaging else 1
+    else:
+      values["CF_Lkas_FcwOpt_USM"] = 2 if enabled else 1
```
Comment (sunnyhaibin) [inline opendbc/car/hyundai/hyundaican.py:57]:
```suggestion
    if mads.enable_mads:
      lkas_icon = 2 if mads.lat_active or mads.disengaging else 1
    else:
      lkas_icon = 2 if enabled else 1
    values["CF_Lkas_FcwOpt_USM"] = lkas_icon
```

File: opendbc/car/nissan/interface.py [hunk 3]
```diff
@@ -23,4 +23,7 @@ def _get_params(ret: structs.CarParams, candidate, fingerprint, car_fw, experime
       # Altima has EPS on C-CAN unlike the others that have it on V-CAN
       ret.safetyConfigs[0].safetyParam |= Panda.FLAG_NISSAN_ALT_EPS_BUS
 
+    if candidate in (CAR.NISSAN_LEAF, CAR.NISSAN_LEAF_IC):
+      ret.safetyConfigs[-1].safetyParam |= Panda.FLAG_NISSAN_LEAF
+
```
Comment (sunnyhaibin) [inline opendbc/car/nissan/interface.py:28]:
```suggestion
    # Used for panda safety and tests
    if candidate in (CAR.NISSAN_LEAF, CAR.NISSAN_LEAF_IC):
      ret.safetyConfigs[-1].safetyParam |= Panda.FLAG_NISSAN_LEAF

```

File: opendbc/car/subaru/carstate.py [hunk 4]
```diff
@@ -7,10 +7,14 @@
 from opendbc.car.subaru.values import DBC, CanBus, SubaruFlags
 from opendbc.car import CanSignalRateCalculator
 
+from opendbc.sunnypilot.car.subaru.mads import MadsCarState
 
-class CarState(CarStateBase):
+ButtonType = structs.CarState.ButtonEvent.Type
+
```
Comment (sunnyhaibin) [inline opendbc/car/subaru/carstate.py:13]:
```suggestion


```

File: opendbc/car/hyundai/carcontroller.py (hunk 1 above)
Comment (sunnyhaibin) [reply opendbc/car/hyundai/carcontroller.py]: No longer in sunnypilot main repo.

File: opendbc/sunnypilot/car/hyundai/escc.py [hunk 5]
```diff
@@ -1,3 +1,29 @@
...
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+
+Last updated: July 29, 2024
+"""
+
```
Comment (sunnyhaibin) [inline opendbc/sunnypilot/car/hyundai/escc.py:26]: Do this in another PR.

```suggestion
```

**Review by sunnyhaibin — APPROVED** [review]
Send it.

**Comment by devtekve** [comment]
> Send it.

Roger that! Will do a final run and merge if all good
